### PR TITLE
Trainer

### PR DESCRIPTION
**Pull request by ms2892**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,4 +4,6 @@ test.ipynb
 .ipynb_checkpoints
 runs/
 __pycache__/
-datasets_debug.log
\ No newline at end of file
+datasets_debug.log
+test*.ipynb
+trainer_debug.log
```
Comment (augustasio) [inline .gitignore]: It's minor, but having the newline would be nice for consistency I think

File: models/trainer.py [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+            if self.val_size>len(self.dataset):
+                
+                # Raise error if val_size is greater than training size 
+                raise Exception("Please give validation size less than training size")
+            
+            # Define the training size
+            self.train_size = len(self.dataset)-self.val_size
+            
+            # Split the dataset into training and validation
+            self.train_dataset,self.val_dataset = torch.utils.data.random_split(self.dataset, [self.train_size, self.val_size])
```
Comment (augustasio) [inline models/trainer.py]: This seems restrictive, what if a user already has the splits and wants to put them in here? I think that might happen for us two. I think just passing in the two datasets separately might be better.

File: models/trainer.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+        else:
+            earlystop = None
+            
+        # Training Loop
+        for step, epoch in enumerate(range(1,self.num_epochs+1)):
+            print('Epoch {}/{}'.format(epoch,self.num_epochs))
+            print('-'*10)
+            
+            for phase in ['train','val']:
+                if (phase=='train'):
```
Comment (augustasio) [inline models/trainer.py:257]: Don't need the brackets

File: models/trainer.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+        # Define train and val datasets
+        dataloaders = {'train':self.train_loader,'val':self.val_loader}
+        dataset_sizes = {'train':len(self.train_dataset),'val':len(self.val_dataset)}
+        
+        # Notify the model name
+        print("The tensorboard model name corresponding to this model is", self.model_name)
+        
+        # Create the early stopping checkpoint class
+        if self.es_flag:
+            earlystop = self.EarlyStop(mode = self.c_flag)
```
Comment (augustasio) [inline models/trainer.py]: EarlyStop does not have a mode parameter in constructor

File: models/trainer.py (hunk 2 above)
Comment (augustasio) [reply models/trainer.py]: Also, the random split is not seeded (unless I missed something?), so one would get a different split each time, not good for reproducibility.

**Review by AugustasMacijauskas — APPROVED** [review]

**Pull request merged**